### DISCUSSION TITLE
How to use uctypes?

### DISCUSSION THREAD

**Discussion by sk8board** [micropython/micropython#15873] [Using MicroPython]
### Example: 
every 100 microseconds, bit 31 is deleted then bit 0 is appended to a 32-bit array. Then key bits are parsed from the 32-bit array.

### Problem:

I am trying to use MicroPython to monitor a bitstream of incoming data to parse key bits as inputs for functions. The bitstream runs at 5000 bit/sec.

I know this would be easier to perform with c, but I want to learn how to do this with MicroPython.

Is there an example or explanation of how to use uctypes with a constant flow of changing bits?

`uctypes` looks to be a perfect tool for monitoring a bitstream of data. The problem is there is very little information on how to implement `uctypes`. 

I have reviewed the three examples in the MicroPython documentation, but these examples were not helpful.

**Comment by andrewleech** [comment]
I'm not too sure about your particular use case, but in general with uctypes I find the test cases for it helpful examples, look for everthing starting with `uctypes_` in https://github.com/micropython/micropython/tree/master/tests/extmod

**Comment by peterhinch** [comment]
If you're using a Pico 2 the 32-bit shift and add of data in a 4-byte bytearray could be done using the inline assembler. Two or three instructions and lightning fast.

**Comment by sk8board** [comment]
@andrewleech Thank you for the link. Unfortunately, the examples were not helpful enough. I decided to try parsing data with brute force using bit shifting.

@peterhinch Thank you for your suggestion. I agree the PIO assembler builds bit arrays with amazing speed. That part of my code is working very well.

My main concern is if MicroPython will be too slow to parse a 32-bit message before being interrupted by the interrupt when the next 32-bit message becomes available. I am now integrating each part of my code together, so I will soon learn if my concern is valid.

**Comment by bixb922** [comment]
You probably have already seen this: code that analyzes 32 bit integer runs very fast if the Viper decorator and Viper data types are used.

**Reply by sk8board** [reply]
Thank you for you input. I did look at it briefly, but the lost interest at the word integer. I am looking for an efficient and quick way to parse bits from a bitstream.

**Reply by bixb922** [reply]
Viper is very efficient for integers and integer arrays. Looking at the OP, this should be a good fit, especially if you are thinking about bit shifting or extracting bits or bit fields.

**Comment by andrewleech** [comment]
I think there's not quite enough information to give full guidance. I can pretty much guarantee that micropython should be fast enough to "process" a 32bit value only 10 times a second, however the detail of what you're trying to process is the make or break.

What I gather so far is you're sampling a bitstream (pin input?) in PIO? such that you have a 32bit buffer which is bit shifted and the new input bit placed at LSB.

Whether you're using PIO or timer interrupt or pin interrupt to do this bit shift and sample possibly doesn't matter.

What might matter more is the format you're getting this 32bit value into python. Is it a 4-long `bytes` or `bytearray` or an `int` value currently?

What are you trying to process? Is it just a bitmask you're trying and `and` against it, or `xor` to pattern match or just looking for a fixed/known 4-byte pattern? Or something more complicated?

Regardless, if you ever need to loop over values to check something bit by bit then viper likely will work well for you.

**Reply by sk8board** [reply]
> What might matter more is the format you're getting this 32bit value into python. Is it a 4-long `bytes` or `bytearray` or an `int` value currently?
> 
> What are you trying to process? Is it just a bitmask you're trying and `and` against it, or `xor` to pattern match or just looking for a fixed/known 4-byte pattern? Or something more complicated?

Here is a snip of my code and a screenshot of PulseView logic analyzer for reference, https://github.com/orgs/micropython/discussions/15897

* For reference, the pulse width is measured using GPIO 16 and State Machine 0.

* State Machine 0 handler determines if the pulse width represents a 0 or 1, then sends the 0 or 1 to State Machine 1.

* State Machine 1 buffers each bit until State Machine 1 TX FIFO is full, then State Machine 1 handler receives the 32-bit array.

![image](https://github.com/user-attachments/assets/6f3e0401-5abd-42ce-ac37-e27da70a0d30)

This provides a 32-bit array that needs to be searched for a preamble. Once the start of the preamble has been determined, then determine the state of a specific bit after the preamble to determine the state of a switch.

example bit array: 0b11111111110101001001011000101110

The preamble consists of 10 ones

A bit of interest is 27 bits behind the first bit of the preamble.

Since I will not know where the preamble is located within the bit array and the message is almost 32-bits long, I will need to concatenate two 32-bit arrays to make a 64-bit array to ensure all of the data is contained in the array.
